File: src/lib/tauri/ai.ts
import { invoke } from "@tauri-apps/api/core";

import type {
  AiInlineSuggestionRequest,
  AiIntentSuggestionRequest,
  AiRecoverySuggestionRequest,
  AiSuggestionCommandResult,
  SuggestionResponse,
} from "../../domain/suggestion/types";
import type {
  AiConnectionTestRequest,
  AiConnectionTestResult,
  CompletionRequest,
  CompletionResponse,
} from "../../domain/ai/types";

export async function requestGhostCompletion(request: CompletionRequest): Promise<CompletionResponse | null> {
  try {
    return await invoke<CompletionResponse | null>("request_completion", { request });
  } catch {
    return null;
  }
}

export async function requestAiInlineSuggestions(
  request: AiInlineSuggestionRequest,
): Promise<AiSuggestionCommandResult | null> {
  try {
    return normalizeSuggestionCommandResult(
      await invoke<AiSuggestionCommandResult | SuggestionResponse | null>("request_ai_inline_suggestions", { request }),
    );
  } catch {
    return {
      status: "networkError",
      suggestions: [],
    };
  }
}

export async function requestAiRecoverySuggestions(
  request: AiRecoverySuggestionRequest,
): Promise<AiSuggestionCommandResult | null> {
  try {
    return normalizeSuggestionCommandResult(
      await invoke<AiSuggestionCommandResult | SuggestionResponse | null>("request_ai_recovery_suggestions", { request }),
    );
  } catch {
    return {
      status: "networkError",
      suggestions: [],
    };
  }
}

export async function requestAiIntentSuggestions(
  request: AiIntentSuggestionRequest,
): Promise<AiSuggestionCommandResult | null> {
  try {
    return normalizeSuggestionCommandResult(
      await invoke<AiSuggestionCommandResult | SuggestionResponse | null>("request_ai_intent_suggestions", { request }),
    );
  } catch {
    return {
      status: "networkError",
      suggestions: [],
    };
  }
}

export async function testAiConnection(request: AiConnectionTestRequest): Promise<AiConnectionTestResult> {
  try {
    return await invoke<AiConnectionTestResult>("test_ai_connection", { request });
  } catch (error) {
    return {
      status: "provider_error",
      message: error instanceof Error ? error.message : String(error),
    };
  }
}

function normalizeSuggestionCommandResult(
  response: AiSuggestionCommandResult | SuggestionResponse | null,
): AiSuggestionCommandResult | null {
  if (!response) {
    return null;
  }

  if ("status" in response) {
    return response;
  }

  return {
    status: response.suggestions.length > 0 ? "success" : "empty",
    suggestions: response.suggestions,
    latencyMs: response.latencyMs,
  };
}

File: src-tauri/src/commands/ai.rs
use crate::ai::{
    build_ai_suggestion_command_result, classify_ai_suggestion_error, complete, inline_suggestions,
    intent_suggestions, recovery_suggestions, test_connection, AiInlineSuggestionRequest,
    AiIntentSuggestionRequest, AiRecoverySuggestionRequest, AiSuggestionCommandResult,
    CompletionRequest, CompletionResponse, ConnectionTestRequest, ConnectionTestResult,
};

#[tauri::command]
pub async fn request_completion(
    request: CompletionRequest,
) -> Result<Option<CompletionResponse>, String> {
    match complete(request).await {
        Ok(response) => Ok(response),
        Err(_) => Ok(None),
    }
}

#[tauri::command]
pub async fn request_ai_inline_suggestions(
    request: AiInlineSuggestionRequest,
) -> Result<AiSuggestionCommandResult, String> {
    match inline_suggestions(request).await {
        Ok(response) => Ok(build_ai_suggestion_command_result(response)),
        Err(error) => Ok(classify_ai_suggestion_error(error)),
    }
}

#[tauri::command]
pub async fn request_ai_recovery_suggestions(
    request: AiRecoverySuggestionRequest,
) -> Result<AiSuggestionCommandResult, String> {
    match recovery_suggestions(request).await {
        Ok(response) => Ok(build_ai_suggestion_command_result(response)),
        Err(error) => Ok(classify_ai_suggestion_error(error)),
    }
}

#[tauri::command]
pub async fn request_ai_intent_suggestions(
    request: AiIntentSuggestionRequest,
) -> Result<AiSuggestionCommandResult, String> {
    match intent_suggestions(request).await {
        Ok(response) => Ok(build_ai_suggestion_command_result(response)),
        Err(error) => Ok(classify_ai_suggestion_error(error)),
    }
}

#[tauri::command]
pub async fn test_ai_connection(
    request: ConnectionTestRequest,
) -> Result<ConnectionTestResult, String> {
    Ok(test_connection(request).await)
}
